File: 2_HuffManCoding.java
import java.util.*;

public class HuffmanCoding {
    private static class Node implements Comparable<Node> {
        char ch;
        int freq;
        Node left, right;

        Node(char c, int f) {
            ch = c;
            freq = f;
            left = right = null;
        }

        @Override
        public int compareTo(Node other) {
            return this.freq - other.freq;
        }
    }

    public void printCodes(Node root, String code) {
        if (root == null)
            return;

        // Leaf node
        if (root.left == null && root.right == null) {
            System.out.println(root.ch + " : " + code);
        }

        printCodes(root.left, code + "0");
        printCodes(root.right, code + "1");
    }

    public void huffmanCoding(char[] ch, int[] freq, int n) {
        PriorityQueue<Node> pq = new PriorityQueue<>();

        // Create leaf nodes
        for (int i = 0; i < n; i++) {
            pq.add(new Node(ch[i], freq[i]));
        }

        // Build Huffman Tree
        while (pq.size() > 1) {
            Node left = pq.poll();
            Node right = pq.poll();

            Node newNode = new Node('$', left.freq + right.freq);
            newNode.left = left;
            newNode.right = right;

            pq.add(newNode);
        }

        // Root of tree
        Node root = pq.poll();

        System.out.println("\nHuffman Codes:");
        printCodes(root, "");
    }

    public static void main(String[] args) {
        HuffmanCoding h = new HuffmanCoding();
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of characters: ");
        int n = sc.nextInt();

        char[] ch = new char[n];
        int[] freq = new int[n];

        System.out.println("\nEnter characters and frequencies:");
        for (int i = 0; i < n; i++) {
            ch[i] = sc.next().charAt(0);
            freq[i] = sc.nextInt();
        }

        h.huffmanCoding(ch, freq, n);

        sc.close();
    }
}

File: 2_HuffManCoding.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Huffman
{
    struct Node
    {
        char ch;
        int freq;
        Node *left, *right;

        Node(char c, int f)
        {
            ch = c;
            freq = f;
            left = right = NULL;
        }
    };

    // Comparator for priority queue
    struct compare
    {
        bool operator()(Node* l, Node* r)
        {
            return l->freq > r->freq;
        }
    };

public:

    // Print Huffman Codes
    void printCodes(Node* root, string code)
    {
        if(root == NULL)
            return;

        // Leaf node
        if(root->left == NULL &&
           root->right == NULL)
        {
            cout << root->ch
                 << " : "
                 << code << endl;
        }

        printCodes(root->left, code + "0");

        printCodes(root->right, code + "1");
    }

    // Huffman Algorithm
    void huffmanCoding(char ch[],
                        int freq[],
                        int n)
    {
        priority_queue<Node*,
                       vector<Node*>,
                       compare> pq;

        // Create leaf nodes
        for(int i = 0; i < n; i++)
        {
            pq.push(new Node(ch[i], freq[i]));
        }

        // Build Huffman Tree
        while(pq.size() != 1)
        {
            Node *left = pq.top();
            pq.pop();

            Node *right = pq.top();
            pq.pop();

            Node *newNode =
            new Node('$',
                     left->freq + right->freq);

            newNode->left = left;
            newNode->right = right;

            pq.push(newNode);
        }

        // Root of tree
        Node *root = pq.top();

        cout << "\nHuffman Codes:\n";

        printCodes(root, "");
    }
};

int main()
{
    Huffman h;

    int n;

    cout << "Enter number of characters: ";
    cin >> n;

    char ch[100];
    int freq[100];

    cout << "\nEnter characters and frequencies:\n";

    for(int i = 0; i < n; i++)
    {
        cin >> ch[i] >> freq[i];
    }

    h.huffmanCoding(ch, freq, n);

    return 0;
}

